base/math.jl

```diff
@@ -1276,14 +1276,14 @@ end
     return ifelse(isfinite(x) & isfinite(err), muladd(x, y, err), x*y)
 end
 
-function ^(x::Float32, n::Integer)
-    n == -2 && return (i=inv(x); i*i)
-    n == 3 && return x*x*x #keep compatibility with literal_pow
-    n < 0 && return Float32(Base.power_by_squaring(inv(Float64(x)),-n))
-    Float32(Base.power_by_squaring(Float64(x),n))
+for (T, WT) in ((Float16, Float32), (Float32, Float64))
+    @eval function ^(x::$T, n::Integer)
+        n == -2 && return (i=inv(x); i*i)
+        n == 3 && return x*x*x #keep compatibility with literal_pow
+        n < 0 && return $T(Base.power_by_squaring(inv($WT(x)),-n))
+        $T(Base.power_by_squaring($WT(x),n))
+    end
 end
```

Comment:
Feel free to ignore but FYI you can do this without the loop and `@eval` like so:
```suggestion
function ^(x::Union{Float16,Float32}, n::Integer)
    n == -2 && return (i=inv(x); i*i)
    n == 3 && return x*x*x #keep compatibility with literal_pow
    n < 0 && return oftype(x, Base.power_by_squaring(inv(widen(x)),-n))
    oftype(x, Base.power_by_squaring(widen(x),n))
end
```
For me this produces the same `code_llvm` as what you've implemented.

Reply:
Ah, nice, thanks, didn't think of using `widen`!